### PR TITLE
Improve dropdown rendering speed

### LINKED ISSUE #10174: Rendering lots of objects in nested templates can be really slow
## Affected Version

4.x

## Description

I have a project with an editable gridfield with many dropdowns (something like 50 rows and 2 dropdowns per row) and it's super slow to render. Here is a screenshot given by the DebugBar in the admin. It takes about 9 seconds to render all the dropdowns (it's faster on production, but still quite slow).

Screenshot 2021-12-10 135247
(the CMSTabSet is also super slow, but that's the next issue i'll be handling :) )

In the meantime, i created this
https://github.com/lekoala/silverstripe-base/blob/master/src/Forms/FastDropdownField.php
which is not really a solution, but at least it works for my use case and make the page much faster where i can prerender the template and avoid calling the view engine each time which seem to be the culprit, because it takes less than 1s now.. should be also be there are also a couple of other dropdowns in other tabs).

## Steps to Reproduce

Create an editable gridfield that is not paginated with many dropdowns. Here is some sample code below, simplified for the purposes of this ticket (not tested or isolated on a fresh project).

```
                $c= GridFieldConfig_RecordEditor::create();
                $c->addComponent(new GridFieldTitleHeader());
                $c->removeComponentsByType(GridFieldSortableHeader::class);
                $c->removeComponentsByType(GridFieldPaginator::class);
                $c->removeComponentsByType(GridFieldPageCount::class);
                $c->removeComponentsByType(GridFieldDeleteAction::class);
                $c->removeComponentsByType(GridFieldAddNewButton::class);
                $gf= new GridField('Records', 'Records', $this->owner->List(), $c);
                $fields->addFieldsToTab('Root.Records', $Records);
        
                $c->removeComponentsByType(GridFieldDataColumns::class);
                $c->addComponent($cols = new GridFieldEditableColumns());
               
                $dataMap= $this->dataMap();
                $fields = [];

                $fields['RelationID'] = [
                    'title' => 'Relation',
                    'callback' => function ($r, $name, $grid) use ($dataMap) {
                        $dd = new DropdownField($name, null, $dataMap, $r->$name);
                        return $dd;
                    }
                ];
                $fields['OtherRelationID'] = [
                    'title' => 'Other Relation',
                    'callback' => function ($r, $name, $grid) use ($dataMap) {
                        $dd = new DropdownField($name, null, $dataMap, $r->$name);
                        return $dd;
                    }
                ];
                $cols->setDisplayFields($fields);
```

## Maybe worth exploring

Simplifying the rendering by not using the view engine to render the option list. I'm going to explore that as well (would probably be cleaner than what i did) and update this issue if I found anything

## PRs

[**kitchen sink CI run**](https://github.com/creative-commoners/recipe-kitchen-sink/actions/runs/17026844314)

Note: This PR doesn't fix the problem, it only makes it _slightly_ less of a problem. Don't close the issue after this PR is merged.

- https://github.com/silverstripe/silverstripe-framework/pull/11826

### PR DESCRIPTION
**Pull request by lekoala**
Fixes https://github.com/silverstripe/silverstripe-framework/issues/10174

### CODE REVIEW

**Review by michalkleiner — CHANGES_REQUESTED** [review]
Failing tests are related, please review.

Apart from that, I'm not personally sure about this change. I get that preparing the list in PHP might be faster, but with that (at a stretch) we can scrape the whole templating system and do everything in PHP.

**Comment by lekoala** [comment]
@michalkleiner yes i'm not sure why the tests are failing i need to review this and see what are the tests doing

the template system is fine for simple things, but not great for loops. rendering large list of items (eg: a dropdown with country names) is quite slow. i do think there is an issue with the templating system itself, but that's clearly out of scope :-)

**Comment by michalkleiner** [comment]
Saw the comment [here](https://github.com/silverstripe/silverstripe-framework/issues/10174#issuecomment-1606894809) so I got more context now. Let's see what other @silverstripe/core-team thinks about the topic.

**Comment by lekoala** [comment]
@michalkleiner tests all green ;-)

File: src/Forms/DropdownField.php [hunk 1]
```diff
@@ -124,6 +126,39 @@ public function getHasEmptyDefault()
         return parent::getHasEmptyDefault() || $this->Required();
     }
 
+    /**
+     * Provides <option> list for the template
+     * This allows for a faster rendering of dropdowns with lots of options
+     *
+     * @return string
+     */
+    public function renderOptionsHTML(): string
```
Comment (michalkleiner) [inline src/Forms/DropdownField.php]: Not sure here but does it need to be public? People can override the `Field` method if needed.
```suggestion
    private function renderOptionsHTML(): string
```

**Comment by GuySartorelli** [comment]
I don't like this change. If there's a problem with the templating system that makes it too slow for this and similar scenarios, then resolving that problem is the fix. This feels like a project-level workaround.

I especially agree with Michal's comment: "I get that preparing the list in PHP might be faster, but with that (at a stretch) we can scrape the whole templating system and do everything in PHP."
If we start pulling template logic into PHP here, where does that end? This doesn't feel like the correct approach to the problem. I'm going to close this PR on that basis.

**Comment by lekoala** [comment]
@GuySartorelli i agree with you: i don't like this change. but when a page takes 2 times more to render because of a couple of dropdowns, that's not something that can be ignored either :-)

ideally, i think profiling should be done on the templating system and see where are the bottlenecks

**Pull request closed without merging**